tardis/energy_input/gamma_ray_packet_source.py

```diff
@@ -0,0 +1,856 @@
+                decay_time_max = self.times[decay_time[i] + 1]
+            # rejection sampling
+            while (decay_times[i] <= decay_time_min) or (
+                decay_times[i] >= decay_time_max
+            ):
+                decay_times[i] = -self.taus[isotope] * np.log(
+                    np.random.random()
+                )
+        return decay_times
+
+    def create_packets(
+        self, decays_per_isotope, number_of_packets, *args, **kwargs
+    ):
+        """Initialize a collection of GXPacket objects for the simulation
+        to operate on.
+
+        Parameters
+        ----------
+        decays_per_isotope : array int64
+            Probability of decays per simulation shell per isotope per time step
+        number_of_packets : int
+            Number of packets to create
+
+        Returns
+        -------
+        GXPacketCollection
+        """
+        # initialize arrays for most packet properties
+        locations = np.zeros((3, number_of_packets))
+        directions = np.zeros((3, number_of_packets))
+        packet_energies_rf = np.zeros(number_of_packets)
+        packet_energies_cmf = np.zeros(number_of_packets)
+        nus_rf = np.zeros(number_of_packets)
+        nus_cmf = np.zeros(number_of_packets)
+        times = np.zeros(number_of_packets)
+        statuses = np.ones(number_of_packets, dtype=np.int64) * 3
+
+        self.energy_plot_positron_rows = np.zeros((number_of_packets, 4))
+
+        # compute positronium continuum
+        positronium_energy, positronium_intensity = positronium_continuum()
+
+        # sample packets from dataframe, returning a dataframe where each row is
+        # a sampled packet
+        sampled_packets_df = decays_per_isotope.sample(
+            n=number_of_packets,
+            weights="decay_energy_erg",
+            replace=True,
+            random_state=np.random.RandomState(self.base_seed),
+        )
+        # get unique isotopes that have produced packets
+        isotopes = pd.unique(sampled_packets_df["isotopes"])
+        # compute the positron fraction for unique isotopes
+        isotope_positron_fraction = self.calculate_positron_fraction(isotopes)
+
+        # get the inner and outer velocity boundaries for each packet to compute
+        # the initial radii
+        sampled_packets_df["inner_velocity"] = self.inner_velocities[
+            sampled_packets_df["shell"].values
+        ]
+        sampled_packets_df["outer_velocity"] = self.outer_velocities[
+            sampled_packets_df["shell"].values
+        ]
+
+        # sample radii at time = 0
+        initial_radii = self.create_packet_radii(sampled_packets_df)
+
+        # packet decay time
+        times = self.create_packet_times_uniform_energy(
+            number_of_packets,
+            sampled_packets_df["isotopes"],
+            sampled_packets_df["time"],
+        )
+
+        # get the time step index of the packets
+        initial_time_indexes = sampled_packets_df["time"]
```

Comment:
`time` is an index of the `data frame`

Reply:
They are actual `times` and not indices.

Reply:
Okay. I think they would be more useful as indices instead of actual times. Times are useful for plotting but not much else in my opinion. I know I am probably referencing the index incorrectly regardless and will fix the code here.